cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java

```diff
@@ -367,4 +375,83 @@ private HttpResponse doRequest(HttpRequest.Builder requestBuilder) throws Datase
     }
     return builder.addHeader(Constants.Security.Headers.USER_ID, userId);
   }
+
+  private static void logThreadDump() {
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Timeout while making request to dataset service. Thread dump follows:");
+      ThreadMXBean bean = ManagementFactory.getThreadMXBean();
```

Comment:
How is this comparing to `Thread.getAllStackTraces`?

Reply:
not sure I copied this from a branch that @albertshau had created earlier. But I believe a stack trace only contains the class and line number of each element; whereas this gives us more information such as the state of the thread, what locks it is holding/waiting on, etc. That would help us identify deadlocks.

Reply:
I think the getAllStackTraces return a Map from Thread to StackElement[], which is kind of similar to this. I was just curious :)